### PR TITLE
SCYLLA-VERSION-GEN: use semver-compatible version

### PR DESCRIPTION
**Pull request by yaronkaikov**
Scylla-operator require Scylla to have semantic versioning. (Ref: https://semver.org/)

Closes: https://github.com/scylladb/scylla/issues/9543

### CODE REVIEW

**Review by fruch — APPROVED** [review]
LGTM

**Review by vponomaryov — APPROVED** [review]

**Review by vponomaryov — CHANGES_REQUESTED** [review]
It must be hyphen (`-`) and not tilde (`~`).

File: SCYLLA-VERSION-GEN [hunk 1]
```diff
@@ -60,7 +60,7 @@ fi
 
 # Default scylla product/version tags
 PRODUCT=scylla
-VERSION=5.1.dev
+VERSION=5.1.0~dev
```
[the remark this answers is not in the document]
Comment (avikivity) [reply SCYLLA-VERSION-GEN:63]: no, tilde is correct. A tilde signifies negative delta (so 5.1.0~dev is before 5.1.0). With a hyphen 5.1.0-dev would compare after 5.1.0.

**Comment by avikivity** [comment]
Dequeued, noticed CI failure.

**Comment by yaronkaikov** [comment]
> Dequeued, noticed CI failure.

@avikivity is running at the moment

**Comment by avikivity** [comment]
> > Dequeued, noticed CI failure.
> 
> @avikivity is running at the moment

I'm not running anywhere

File: SCYLLA-VERSION-GEN (hunk 1 above)
[the remark this answers is not in the document]
Comment (vponomaryov) [reply SCYLLA-VERSION-GEN:63]: From the `https://semver.org/`
```
<valid semver> ::= <version core>
                 | <version core> "-" <pre-release>
                 | <version core> "+" <build>
                 | <version core> "-" <pre-release> "+" <build>
```

[the remark this answers is not in the document]
Comment (yaronkaikov) [reply SCYLLA-VERSION-GEN:63]: @vponomaryov @avikivity possible solution is to use `5.1.0+dev` , this works for both python and based on https://semvercompare.azurewebsites.net/?version=5.1.0%252Bdev&version=5.1.0 it is also good for packaging

[the remark this answers is not in the document]
Comment (yaronkaikov) [reply SCYLLA-VERSION-GEN:63]: `[yaronkaikov@london]~/git/scylla-operator (master)$ python
Python 3.10.5 (main, Jun  9 2022, 00:00:00) [GCC 12.1.1 20220507 (Red Hat 12.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import semver
>>> semver.parse('5.1.0+dev')
OrderedDict([('major', 5), ('minor', 1), ('patch', 0), ('prerelease', None), ('build', 'dev')])
>>> quit()
`

[the remark this answers is not in the document]
Comment (yaronkaikov) [reply SCYLLA-VERSION-GEN:63]:
```
[yaronkaikov@london]~/git/scylla-operator (master)$ python
Python 3.10.5 (main, Jun  9 2022, 00:00:00) [GCC 12.1.1 20220507 (Red Hat 12.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import semver
>>> semver.parse('5.1.0+dev')
OrderedDict([('major', 5), ('minor', 1), ('patch', 0), ('prerelease', None), ('build', 'dev')])
>>> quit()
```

[the remark this answers is not in the document]
Comment (vponomaryov) [reply SCYLLA-VERSION-GEN:63]: Possible solution? `Hyphen` shows "pre-release" part and `plus` shows "build" part as it is mentioned in the previous comment.
`semver` has strict structure describe on the `https://semver.org/` site.

So, trying to make Scylla versions be `semver`-like we must comply exactly with `semver` structure and not any other.
As of now the commit's goal which is `use semver-compatible version` is not addressed.

[the remark this answers is not in the document]
Comment (yaronkaikov) [reply SCYLLA-VERSION-GEN:63]: @avikivity based on https://semvercompare.azurewebsites.net/?version=5.1.0-dev&version=5.1.0 , 5.1.0 will come **after** 5.1.0-dev which is good. the problem is that `rpmbuild` failed to build the package when the version contains `-` (debian works fine)

[the remark this answers is not in the document]
Comment (yaronkaikov) [reply SCYLLA-VERSION-GEN:63]: @syuu1228 /cc

**Comment by yaronkaikov** [comment]
@avikivity @vponomaryov Please re-review

**Review by vponomaryov — APPROVED** [review]
LGTM

**Comment by yaronkaikov** [comment]
@avikivity ping

File: SCYLLA-VERSION-GEN [hunk 2]
```diff
@@ -60,7 +60,7 @@ fi
 
 # Default scylla product/version tags
 PRODUCT=scylla
-VERSION=5.1.dev
+VERSION=5.1.0-dev
```
Comment (avikivity) [inline SCYLLA-VERSION-GEN:63]: A dash separates the version from the release. We should use ~ here so it's part of the version.

File: SCYLLA-VERSION-GEN [hunk 3]
```diff
@@ -69,13 +69,18 @@ then
 else
 	DATE=$(date --utc +%Y%m%d)
 	GIT_COMMIT=$(git -C "$SCRIPT_DIR" log --pretty=format:'%h' -n 1 --abbrev=12)
-	SCYLLA_VERSION=$VERSION
+	SCYLLA_VERSION=$(echo $VERSION | awk -F'-' '{print $1}')
+	RELEASE_CANDIDATE=$(echo $VERSION | awk -F'-' '{print $2}')
```
Comment (avikivity) [inline SCYLLA-VERSION-GEN:73]: Why pile more and more complexity into this thing.

File: SCYLLA-VERSION-GEN (hunk 2 above)
Comment (yaronkaikov) [reply SCYLLA-VERSION-GEN:63]: @avikivity the problem is that `~` is not compatible with semantic version

```
[yaronkaikov@london]~/git/scylla-pkg (run-cloud-image-build-from-packer-dir)$ python
Python 3.10.5 (main, Jun  9 2022, 00:00:00) [GCC 12.1.1 20220507 (Red Hat 12.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import semver
>>> semver.parse('5.1.0~dev')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/site-packages/semver.py", line 159, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib/python3.10/site-packages/semver.py", line 190, in parse
    return VersionInfo.parse(version).to_dict()
  File "/usr/lib/python3.10/site-packages/semver.py", line 726, in parse
    raise ValueError("%s is not valid SemVer string" % version)
ValueError: 5.1.0~dev is not valid SemVer string
>>> semver.parse('5.1.0-dev')
OrderedDict([('major', 5), ('minor', 1), ('patch', 0), ('prerelease', 'dev'), ('build', None)])
>>> semver.parse('5.1.0')
OrderedDict([('major', 5), ('minor', 1), ('patch', 0), ('prerelease', None), ('build', None)])

```

File: SCYLLA-VERSION-GEN (hunk 3 above)
Comment (yaronkaikov) [reply SCYLLA-VERSION-GEN:73]: When we are building rpm packages the spec file can't have `-` as part of either `version` or `release` parameters. So what we need is to add the `dev` or in releases `rcX` as part of the `release` parameter
For ubuntu it's working

File: SCYLLA-VERSION-GEN (hunk 2 above)
Comment (avikivity) [reply SCYLLA-VERSION-GEN:63]: I see. What does scylla-operator use for the semver name? The rpm/deb package name? The output of scylla --version?

**Comment by avikivity** [comment]
The problem is discussed in https://github.com/semver/semver/issues/145.

Here's a proposal:

1. define VERSION as in semver (e.g. 5.1.0-rc1)
2. keep SCYLLA_RELEASE as is
3. for rpm/deb, use ${VERSION/-/"~"} for the version field, to conform to rpm/deb expectations
4. for everything else (relocatable/docker) use VERSION as is

File: SCYLLA-VERSION-GEN (hunk 3 above)
Comment (avikivity) [reply SCYLLA-VERSION-GEN:73]: rpm/deb define the version field as the software version, with release only about post-packaging changes. So pushing rc there is not correct.

**Comment by yaronkaikov** [comment]
@avikivity ci passed , please review

**Comment by yaronkaikov** [comment]
@avikivity ping review

**Comment by avikivity** [comment]
Make sure to update the branching scripts and release candidate scripts to update the version instead of the release.

**Pull request closed without merging**